### PR TITLE
feat: add support for pattern matching

### LINKED ISSUE #116: Support pattern matching in python 3.10
[PEP 636](https://www.python.org/dev/peps/pep-0636/)

### PR DESCRIPTION
**Pull request by theHamsta**
https://www.python.org/dev/peps/pep-0636/

Fixes #116

### CODE REVIEW

**Comment by theHamsta** [comment]
I need to solve this somehow with patterns (including "or" and "as" patterns)

File: grammar.js [hunk 1]
```diff
@@ -522,6 +534,14 @@ module.exports = grammar({
       choice($.identifier, $.keyword_identifier, $.subscript, $.attribute)
     ),
 
+    // Extended patterns (patterns allowed in match statement are far more flexible than simple patterns
+
+    as_pattern: $ => prec.left(seq(
```
Comment (theHamsta) [inline grammar.js:539]: or call this alias_expression?

**Comment by theHamsta** [comment]
Test fail for uses of match not as a keyword
```
match = 4
```

**Comment by theHamsta** [comment]
A possible hacky resolution would be to have a common parent for assignments that assign to `"match"` and match_statement
```
        maybe_match_statement: $ => seq(
            choice($.match_statement,
                alias(seq(alias("match", $.identifier),
                    choice(
                        seq('=', field('right', $._right_hand_side)),
                        seq(':', field('type', $.type)),
                        seq(':', field('type', $.type), '=', field('right', $._right_hand_side))
                    )), $.assignment))),
```
I can't make  `maybe_match_statement` hidden due to super node requirement in compound statement.

File: grammar.js [hunk 2]
```diff
@@ -253,6 +257,24 @@ module.exports = grammar({
...
     ),
 
+    _maybe_match_statement: $ => seq(
+      choice($.match_statement,
+        alias(seq(alias("match", $.identifier),
+          choice(
+            seq('=', field('right', $._right_hand_side)),
+            seq(':', field('type', $.type)),
+            seq(':', field('type', $.type), '=', field('right', $._right_hand_side))
+          )), $.assignment))),
```
Comment (theHamsta) [inline grammar.js:267]: hack to allow `match = 3`. Need to find a better solution!

**Comment by thomkeh** [comment]
Maybe there could also be a test case for a variable called `match` with a type annotation:

```python
match: int = 3
```

**Comment by theHamsta** [comment]
@thomkeh sure, but right now it doesn't parse the simple case correctly

File: grammar.js (hunk 1 above)
Comment (theHamsta) [reply grammar.js:545]: The patterns in a match case are a subset of expressions. I simplified using expression here

File: grammar.js [hunk 3]
```diff
@@ -537,12 +554,14 @@ module.exports = grammar({
...
       $.conditional_expression,
-      $.named_expression
+      $.named_expression,
+      $.as_pattern
     ),
 
     primary_expression: $ => choice(
       $.binary_operator,
       $.identifier,
+      alias("match", $.identifier),
```
Comment (theHamsta) [inline grammar.js:564]: just adding match to "keyword" identifiers did make it work for me to parse `a = [match]` correctly

**Comment by GBeauregard** [comment]
if you're looking for weird match test cases you can go through the ones in CPython: https://github.com/python/cpython/blob/main/Lib/test/test_patma.py

There's plenty of weird ones https://github.com/python/cpython/blob/eb483c46d62707bdf705491f76cf1fa9642fb47e/Lib/test/test_patma.py#L1205

**Comment by theHamsta** [comment]
Thanks @GBeauregard I hope I fixed a few edge cases in test_patma.py

**Comment by GBeauregard** [comment]
> Test fail for uses of match not as a keyword (django loves match)
> 
> ```
> match = 4
> ```
> 
> Match is a soft keyword, the python parser would backtrack when seeing a `:` and start parsing a match statement https://www.python.org/dev/peps/pep-0622/#id69

I think you figured it out (and I don't think it mattered for this comment anyway), but to clarify the PEP 622 you linked here is not the accepted pattern matching specification (note the status is superseded); the final/accepted specification differs in a few aspects and was put in PEP 634: https://www.python.org/dev/peps/pep-0634/

> Thanks @GBeauregard I hope I fixed a few edge cases in test_patma.py . The file now parses successfully.

no problem, thanks for implementing parsing for the fun new feature :)

**Comment by GBeauregard** [comment]
One more comment, this and the `nvim-treesitter` PR seem to missing the new `_` soft keyword. See https://docs.python.org/3.10/reference/lexical_analysis.html#soft-keywords for reference on it being considered a keyword and https://github.com/python/cpython/pull/25851 where the CPython built-in IDLE highlighting considers and highlights this as a soft keyword.

This can be useful for users as well, since the syntax highlighting difference will point out that `_` actually has a different behavior from a capture variable named `_` (can be used multiple times, etc etc).

**Comment by theHamsta** [comment]
This PR just regards all of the patterns introduced by pattern matching as expressions. To treat `_` special only in patterns one would have to implement all the introduced patterns. Not sure whether that's worth it since `_` is only semantically different from an identifier (it can bind to multiple values). For highlighting, one could just always highlight `_` as wildcards.

**Comment by bellini666** [comment]
Hey guys. Just wondering when this will get merged :)

**Review by dcreager — COMMENTED** [review]
This looks great, @theHamsta, thanks for taking this on!

You'll need to rebase onto the latest master since I just merged a couple of other PRs.  Also, can you add `match` and `case` as new keywords in the syntax highlighting file?  And ideally some highlighting tests to `test/highlight/keywords.py`.  Then I think this is good to merge!

**Comment by theHamsta** [comment]
@dcreager rebased and added a highlight test

**Review by dcreager — APPROVED** [review]
